[quant][graphmode] Returning axis from calculate_qparams

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29494 [quant][graphmode] Returning axis from calculate_qparams**
* #29493 [quant][graphmode] Test for per channel graph mode quantization
* #29416 [quant] Make PerChannelMinMaxObserver scriptable using `torch.jit.ignore`

Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18580905](https://our.internmc.facebook.com/intern/diff/D18580905)

jerryzh168 committed:
[quant][graphmode] Returning axis from calculate_qparams

Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -134,7 +134,7 @@ def _calculate_per_channel_qparams(self, min_vals, max_vals):
                                     Returning default scale and zero point "
             )
-            return torch.tensor([1.0]), torch.tensor([0])
+            return torch.tensor([1.0]), torch.tensor([0]), self.ch_axis
```
Not sure I follow the reasoning here: The axis is passed as an attribute to observer and is used when statistics are collected. CalcQparams should be independent of the axis argument, it should know if per channel or per tensor calculation is desired, but operations are independent of the axis argument.

jerryzh168 replied:
we don't need axis for calculate_qparams, but we need the axis argument for per channel quantize op. so it's an output value, not an input argument to calculate_qparams.

jerryzh168 replied:
we'll have a separate `get_qparam` function that returns these things, I'll do this in a separate PR

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -403,8 +403,8 @@ Node* insertQuantDeQuantCall(
     zero_point_val = g->insertConstant(zero_point);
-    // TODO: get axis from qparam
-    axis_val = g->insertConstant(0);
+    int axis = tp->elements()[2].toInt();
```
This should be squashed with the change where it was introduced before landing. There is no reason to commit half-finished stuff.

jerryzh168 replied:
this is just because the order we do things right now, at the place we introduce this, calculate_qparam doesn't return axis

ZolotukhinM replied:
Ok, I reread everything more carefully and it makes sense.

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18580905](https://our.internmc.facebook.com/intern/diff/D18580905)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Returning axis from calculate_qparams"


Summary:
`calculate_qparams` of per channel quantization should return the axis, this
PR added this and also added corresponding support in graph mode

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18580905](https://our.internmc.facebook.com/intern/diff/D18580905)

[ghstack-poisoned]
